### ISSUE TITLE
Add RelaxReparam to convert O.H.C. to RelaxedO.H.C.StraightThrough

### ISSUE THREAD

**Issue by fritzo** [pyro-ppl/pyro#2689]
It would help to separate modeling from inference if we allowed model transformations from discrete to relaxed discrete.

One way to accomplish this would be via a poutine, or more specifically via a reparametrizer. For example we could replace
```diff
- x = pyro.sample("x", dist.RelaxedOneHotCategoricalStraightThrough(probs=probs))
+ with poutine.reparam(config={"x": RelaxReparam(temperature=0.5)}):
+     x = pyro.sample("x", dist.OneHotCategorical(probs=probs))
```
## Design questions
There are many ways to relax.
1. Should we be opinionated or flexible, regarding how to relax?
2. How should flexibility be presented?

**Comment by karalets** [comment]
I'd love to have this option, what you are describing feels very pyro-nic
and elegant, since we typically only want to relax models for the sake of
optimization instead of making the modeling choice of relaxedCategorical
etc. !

But it would be nice to keep the 'hard-coded' variants alive for various
reasons, historic and practical.

Would you consider thinking about this more generally also in the context
of softmax tricks ( https://arxiv.org/abs/2006.08063 ) ?

**Comment by eb8680** [comment]
> Should we be opinionated or flexible, regarding how to relax?

What are all of the distributions currently in Pyro and PyTorch we might want to relax? Is there a canonical Gumbel-softmax relaxation for all (or most) of them? 

> How should flexibility be presented?

I guess one thing we could do that would fit with a canonical relaxation approach is to add a `singledispatchmethod` `.relax: Distribution -> Distribution` to a `RelaxReparam` reparametrizer and register canonical relaxations for the built-in distributions. If someone wanted to override them, they could subclass the relevant distribution and register a new `singledispatch` function.  That would slightly blur the model/inference boundary, but hopefully only locally to particular sample sites and only in unusual cases.
